src/chainstate/coordinator/mod.rs

```diff
@@ -0,0 +1,363 @@
+            let current_block = self.burnchain_blocks_db.get_burnchain_block(&cursor)
+                .map_err(|e| {
+                    warn!("ChainsCoordinator: could not retrieve  block burnhash={}", &cursor);
+                    Error::NonContiguousBurnchainBlock(e)
+                })?;
+
+            let parent = current_block.header.parent_block_hash.clone();
+            sortitions_to_process.push_front(current_block);
+            cursor = parent;
+        }
+
+        for unprocessed_block in sortitions_to_process.drain(..) {
+            let BurnchainBlockData { header, ops } = unprocessed_block;
+
+            let sortition_tip_snapshot = SortitionDB::get_block_snapshot(
+                self.sortition_db.conn(), &canonical_sortition_tip)?
+                .expect("BUG: no data for sortition");
+
+            // at this point, we need to figure out if the sortition we are
+            //  about to process is the first block in reward cycle.
+            let reward_cycle_info = self.get_reward_cycle_info(&header);
+            let sortition_id = self.sortition_db.evaluate_sortition(
+                &header, ops, &self.burnchain, &canonical_sortition_tip, reward_cycle_info)
+                .map_err(|e| {
+                    error!("ChainsCoordinator: unable to evaluate sortition {:?}", e);
+                    Error::FailedToProcessSortition(e)
+                })?
+                .0.sortition_id;
+
+            self.notifier.notify_sortition_processed();
+
+            debug!("Sortition processed: {}", &sortition_id);
+
+            if sortition_tip_snapshot.block_height < header.block_height {
+                // bump canonical sortition...
+                self.canonical_sortition_tip = Some(sortition_id.clone());
+                canonical_sortition_tip = sortition_id;
+            }
+
+            if let Some(pox_anchor) = self.process_ready_blocks()? {
+                return self.process_new_pox_anchor(&pox_anchor)
+            }
+        }
+
+        Ok(())
+    }
+
+    /// returns None if this burnchain block is _not_ the start of a reward cycle
+    ///         otherwise, returns the required reward cycle info for this burnchain block
+    ///                     in our current sortition view:
+    ///           * PoX anchor block
+    ///           * Was PoX anchor block known?
+    fn get_reward_cycle_info(&self, burn_header: &BurnchainBlockHeader) -> Option<RewardCycleInfo> {
+        if self.burnchain.is_reward_cycle_start(burn_header.block_height) {
+            info!("Beginning reward cycle. block_height={}", burn_header.block_height);
+            Some(RewardCycleInfo {
+                anchor_block: None,
+                anchor_block_known: true
+            })
+        } else {
+            None
+        }
+    }
+
+    ///
+    /// Process any ready staging blocks until there are no more to process
+    ///  _or_ a PoX anchor block is discovered.
```

Comment:
How do we know which has happened when this method returns?

Reply:
It returns Some(...) if an anchor block is discovered, otherwise returns None.

Reply:
I guess my concern is, how does the caller tell the difference between "there is no anchor block selected for this reward set, and there never will be" versus "there is no anchor block selected, but there might be later once we have all the blocks for the registration window"?  Maybe that's not something the caller needs to know?

Reply:
Right -- the caller here just needs to know if it needs to do an unwind -- I'll update this rustdoc, the behavior here is really:

* Process an ready staging blocks until there are either:
  * no more to process
  * a PoX anchor block is processed which invalidates the current PoX fork

In the normal case that a PoX anchor block is selected, and the node already had processed it, this function would just keep processing (because the caller doesn't need to unwind).

Reply:
Updated in 2f9467106